### ISSUE TITLE
[istio.io testing] Need test for Provisioning Identity through SDS

### ISSUE THREAD

**Issue by brian-avery** [istio/istio#20164]
**Describe the feature request**
We need to develop a test for the [Provisioning Identity through SDS](https://preliminary.istio.io/docs/tasks/security/citadel-config/auth-sds/) task in order to provide automated testing of the Istio.io website for future releases.

Example tests and the documentation for the framework can be found here: https://github.com/istio/istio/tree/master/tests/integration/istioio

A YouTube video describing the usage of the framework can be found here: https://www.youtube.com/watch?v=3y-z8NaVwr0

Please reach out to the Test and Release channel on the [Istio Slack](https://discuss.istio.io/t/istio-slack-channel/1527) or [Test and Release Discuss](https://discuss.istio.io/c/test-and-release)  with any questions.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ x] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ x] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**

**Comment by su225** [comment]
I'll take this one.

**Comment by brian-avery** [comment]
Thanks @su225

**Comment by Pawan-Bishnoi** [comment]
@su225 Have you started working on this. I would like to help here (or just follow along if you don't want help)
Looking for a place to start. Thanks!

**Comment by su225** [comment]
@Pawan-Bishnoi  - Yes. I have started working on this. There are lots of [istio.io testing] issues here. You can pick any of them and start working.

**Comment by su225** [comment]
@brian-avery - Some issues regarding environment setup for SDS. I'm blocked. I dug through a lot, but keep finding new feature gates to enable. I found 4 so far. So any clue how to fix this one? This is blocking my way.

**Here is what I did so far**
I compared the default/demo profile with **sds** profile of Istio with `istio profile diff` and set those Helm values
```go
func enableSDSIdProvisioning(cfg *istio.Config) {
	if cfg == nil {
		return
	}
	cfg.Values["global.sds.enabled"] = "true"
	cfg.Values["global.sds.udsPath"] = "unix:/var/run/sds/uds_path"
	cfg.Values["global.sds.token.aud"] = "istio-ca"

	cfg.Values["global.mtls.enabled"] = "true"

	cfg.Values["nodeagent.enabled"] = "true"
	cfg.Values["nodeagent.image"] = "node-agent-k8s"
	cfg.Values["nodeagent.env.CA_PROVIDER"] = "Citadel"
	cfg.Values["nodeagent.env.CA_ADDR"] = "istio-citadel.istio-system:8060"
	cfg.Values["nodeagent.env.VALID_TOKEN"] = "true"

	cfg.Values["security.components.nodeAgent.enabled"] = "true"
}
```
Pilot, Ingress-gateway and others were stuck in `ContainerCreating` state forever. Here is the error I found
```
Unable to attach or mount volumes: unmounted volumes=[istio-token], 
unattached volumes=[istio-pilot-service-account-token-xdzjz sds-uds-path 
istio-token config-volume istio-certs]: timed out waiting for the condition
```
Even before that it complained that **TokenEndpoint** feature gate was not enabled. So i enabled them by passing `--feature-gates=TokenEndpoint=true,TokenEndpointProjection=true` flags to `minikube start`. I can also see them passed to `kube-apiserver`. In spite of it **no luck**
```diff
```diff
kube-apiserver
  --advertise-address=192.168.0.122
  --allow-privileged=true
  --authorization-mode=Node,RBAC
  --client-ca-file=/var/lib/minikube/certs/ca.crt
  --enable-admission-plugins=NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota
  --enable-bootstrap-token-auth=true
  --etcd-cafile=/var/lib/minikube/certs/etcd/ca.crt
  --etcd-certfile=/var/lib/minikube/certs/apiserver-etcd-client.crt
  --etcd-keyfile=/var/lib/minikube/certs/apiserver-etcd-client.key
  --etcd-servers=https://127.0.0.1:2379
+ --feature-gates=TokenRequest=true,TokenRequestProjection=true
  --insecure-port=0
  --kubelet-client-certificate=/var/lib/minikube/certs/apiserver-kubelet-client.crt
  --kubelet-client-key=/var/lib/minikube/certs/apiserver-kubelet-client.key
  --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
  --proxy-client-cert-file=/var/lib/minikube/certs/front-proxy-client.crt
  --proxy-client-key-file=/var/lib/minikube/certs/front-proxy-client.key
  --requestheader-allowed-names=front-proxy-client
  --requestheader-client-ca-file=/var/lib/minikube/certs/front-proxy-ca.crt
  --requestheader-extra-headers-prefix=X-Remote-Extra-
  --requestheader-group-headers=X-Remote-Group
  --requestheader-username-headers=X-Remote-User
  --secure-port=8443
  --service-account-key-file=/var/lib/minikube/certs/sa.pub
  --service-cluster-ip-range=10.96.0.0/12
  --tls-cert-file=/var/lib/minikube/certs/apiserver.crt
  --tls-private-key-file=/var/lib/minikube/certs/apiserver.key
```
I then started examining the secret on which projection is being done. An example JWT looks like this
```json
{
  "alg": "RS256",
  "kid": "yo-oRLZMTR6xrPyE8cFCuAiMiSVwssAgiF6YbgRAs2M"
}
{
  "iss": "kubernetes/serviceaccount",
  "kubernetes.io/serviceaccount/namespace": "istio-system",
  "kubernetes.io/serviceaccount/secret.name": "istio-pilot-service-account-token-stxdp",
  "kubernetes.io/serviceaccount/service-account.name": "istio-pilot-service-account",
  "kubernetes.io/serviceaccount/service-account.uid": "ce257494-652f-4f04-b4ac-e85dffc9a66e",
  "sub": "system:serviceaccount:istio-system:istio-pilot-service-account"
}
```
It does not look like what the projection YAML expects

But the projection YAML for pilot looks like this
```yaml
- name: istio-token
  projected:
    sources:
    - serviceAccountToken:
        audience: istio-ca
        expirationSeconds: 43200
        path: istio-token
```
So I started digging into **"service account token projection"** and Google took me to Kubernetes documentation page [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection)

To summarize, it mentions one more feature gate called `ServiceAccountTokenVolumeProjection` to be enabled. Besides that, it also mentions some more parameter to be passed to `kube-apiserver` namely
* `--service-account-issuer`
* `--service-account-signing-key-file`
* `--service-account-api-audiences`

There is yet another link to [Istio blog](https://istio.io/blog/2019/trustworthy-jwt-sds/) titled **Change in Secret Discovery Service in Istio 1.3** which talks about introducing audience fields to JWT tokens to limit its scope. I see that this is biting me. While digging further in that blog, I ran into another feature gate named `BoundServiceAccountTokenVolume` which is still in alpha and is off by default.

**Comment by su225** [comment]
Ah! Found something here - https://github.com/istio/istio/blob/master/prow/config/trustworthy-jwt.yaml
Let me try and come back to you.

**Comment by su225** [comment]
Finally it worked. I had to add the following extra configuration to Minikube to make it work
```bash
--extra-config=apiserver.service-account-signing-key-file="/var/lib/minikube/certs/sa.key" \
--extra-config=apiserver.service-account-issuer="kubernetes.default.svc" \
--extra-config=kubelet.resolv-conf="/run/systemd/resolve/resolv.conf"
```
I guess for most people only the first two are relevant. My CoreDNS was stuck in forwarding loop and hence the third one.
